Use SWC loader

tomdracz opened the pull request:
Investigates how can we use `swc-loader` as a replacement for `babel-loader`

So far, works fairly well but need to figure out configuration layer. Things like React support, TS, fast refresh where previously enabled by a check whether the relevant plugin is installed. With swc, they come built in but need to be enabled in the config so might need to use additional configuration layer or settle on some defaults

To explore/to decide/To do:

- [x] Configuration - inline with loader setup or using `.swcrc` file?
- [x] Configuring features - allow running React, TS, fast refresh by default or make user configurable? Is there a reason why we would not enable all of those?
- [x] Installing peer dependencies, no point in installing everything so might need to rejig how we install peer dependencies. For now could maybe leave `swc` out of peer dependencies and have manual install instructions
- [x] Explore how `swc` env matches up with `@babel/preset-env` - some options map one to one, but others are absent (works but `browserslist` is not processed, see https://github.com/swc-project/swc/issues/3365)
- [x] Untested, need to see how it behaves with `.erb` or `.coffee` files
- [x] Cleanup duplication and streamline `swc` loader rules
- [ ] What tests would we cover this with?

tomdracz commented:
Progress so far, tried hooking this up to https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh and got it to work nicely after some hurdles with server rendering.

Browserlists doesn't seem to be correctly applied unless specified directly in swc config. Opened https://github.com/swc-project/swc/issues/3365 to in swc repo.

justin808 commented:
@tomdracz please make a PR for https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh!

EXCITING!

tomdracz commented on package/rules/swc.js:
```diff
@@ -0,0 +1,52 @@
+  exclude: /node_modules/,
+  use: ({ resource }) => ({
+      loader: require.resolve('swc-loader'),
+      options: {
```
Need to decide how to best configure the loader options as they are not easily modifiable by users and config here doesn't support stuff like TS decorators (but SWC itself does!).

Three ways for it I can see:
- use `.swcrc` config instead https://swc.rs/docs/configuration/swcrc - could pull this in on shakapacker install. Biggest issue I see is support fast refresh. We need to load it conditionally as per `refresh: runningWebpackDevServer && isDevelopment`. Without dev server check, SSR compilation struggles with undefined variables
- Enable more features out of the box. We can support top level await, decorators etc by default. Might be better on the opt-out basis though
- Provide additional configuration layer through a config file. We could set the defaults as they are here and enable users to edit some `swc.loader.config.js` file to merge their changes with the defaults. Provides more overhead but it does make it easiest to configure on demand

saiqulhaq replied:
for the first way, you said that it will be an issue based on the known limitation list

I think the third one is a good idea, which is similar to `babel.config.js`, and we could add documentation like `customizing_babel_config.md` doc
we could mention to restart webpack dev server if required on the doc

tomdracz commented on package/utils/helpers.js:
```diff
@@ -39,6 +39,22 @@ const canProcess = (rule, fn) => {
+
+  if (!moduleExists(loaderName)) {
+    throw new Error(
+      `Your webpacker config specified using ${configLoader}, but ${loaderName} package is not installed. Please install ${loaderName} first.`
```
Throw an error when loader specified in `webpacker.yml` has got no corresponding package installed

justin808 replied:
Can we be more specific in that only babel and swc are supported in the error message?

tomdracz commented on package/rules/index.js:
```diff
@@ -15,4 +16,4 @@ const rules = {
 module.exports = Object.keys(rules)
   .filter((key) => !!rules[key])
-  .map((key) => rules[key])
+  .flatMap((key) => rules[key])
```
Can probably revert back to `map` now that `swc.js` exports a single rule

justin808 commented:
@tomdracz let me know when you think this is ready!

tomdracz commented:
@justin808 Think this one is good to go.

I've added configuration layer using `config/swc.config.js` file. This is similar to how you would configure babel although bit more opinionated and specific to us. Still probably best of the options to allow customisation.

https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh/pull/30 also updated to show the migration to `swc-loader`

it will still be experimental until we get more usage (tested on the example app and our current prod app) but it's a start.

Will be good to add generator option to install that would set correct `webpack_loader` option, install the dependencies and maybe set some example `config/swc.config.js` but we can bring this in later when we move this out of "experimental" phase.

justin808 commented:
@tomdracz I will check this out next week for sure, at the latest. Very exciting!

In terms of 

> What tests would we cover this with?

I think the gem should 2 big improvements to CI:

1. A sample app, like `/spec/dummy` https://github.com/shakacode/react_on_rails/tree/master/spec/dummy
2. CI steps that run `rails new` and the webpacker generator, and then a test, like [those here in React on Rails](https://github.com/shakacode/react_on_rails/tree/master/rakelib)

Take a look at https://github.com/shakacode/react_on_rails/blob/master/Rakefile and https://github.com/shakacode/react_on_rails/blob/master/.github/workflows/main.yml.

justin808 commented:
@tomdracz Let me know if you'd like me to push out a pre-release 6.1 with support for this feature.

justin808 commented on docs/using_swc_loader.md:
```diff
@@ -0,0 +1,126 @@
+
+### React
+
+React is supported out of the box, provided you use `.jsx` or `.tsx` file extension. Shakapacker config will correctly recognise those and tell SWC to parse the JSX syntax correctly
```
Can you add periods at the end of sentences?

Also, should we standardize on USA spelling (recognize)?

justin808 commented on docs/using_swc_loader.md:
````diff
@@ -0,0 +1,126 @@
+module.exports = customConfig
+```
+
+### Example: Enabling React Fast Refresh
````
We can either change this package or make a new one that makes this integration super easy.

https://github.com/shakacode/react-on-webpacker/

justin808 commented on package/swc/index.js:
```diff
@@ -0,0 +1,57 @@
+          syntax: isTypescriptFile(filenameToProcess)
+            ? 'typescript'
+            : 'ecmascript',
+          [isTypescriptFile(filenameToProcess) ? 'tsx' : 'jsx']:
```
Should we move React support back into Shakapacker? Seems that we should keep it the same for both babel and swc? However, now that I'm maintaining shakapacker, I don't mind having React support being supported by default.

https://github.com/rails/webpacker/pull/3224

tomdracz replied:
@justin808 there's two parts to it  - understanding jsx syntax and React transform. I think we can move the latter to a config but with the former relying on the filename it might need to stay.

I'll see how it handles non-React jsx files and adjust accordingly.

tomdracz replied:
@justin808 There's some weirdness with passing options to swc inline, for more details https://github.com/swc-project/swc/discussions/3374

This setup will work though regardless and I see no real reason not to support basic React config out of the box if we're not requiring any specific presets or plugins like `@babel/preset-react`. It "just" works so we might as well embrace it. We could drop the tsx/jsx setting here and move to the custom config but that will probably need to become a function that takes a filename to allow dynamic setting of jsx/tsx value (no idea why SWC just didn't settle for one setting 🤷 ).

Still, I've removed the transform setting from the default config and added section to the docs on how might one add settings to match any settings they have in preset-react if they wish to do so.

justin808 replied:
We'll iterate on this after the first release.

Maybe we can just be clear that SWC is experimental and the final API might change by the time 6.1.0 is released.

justin808 commented on README.md:
```diff
@@ -371,6 +372,12 @@ By default, you will find the Webpacker preset in your `package.json`. Note, you
+
+You can try out experimental integration with SWC loader. You can read more at [SWC usage docs](./docs/using_swc_loader.md)
+
+Please note that if you want opt-in to use SWC, you can skip [React](#react) integration instructions as it is supported out of the box
```
See my note below on React.

justin808 reviewed (changes requested):
Looks GREAT!

Just some grammar and spelling is all that I can see without trying this out.

justin808 commented:
@tomdracz once I merge, I'll push a 6.0.0.beta.1 with this change.

justin808 reviewed (approved):
Thanks for cleaning up the grammar!

justin808 commented:
@tomdracz shakapacker-6.1.0.beta.0.gem and npm pushed.

Can you please update the changelog?

💥
